source/agora/network/RPC.d

```diff
@@ -33,12 +33,91 @@ import vibe.core.connectionpool;
 import vibe.core.sync;
 import vibe.http.server : RejectConnectionPredicate;
 
+import std.conv : to;
+import std.socket : Address, AddressFamily, parseAddress;
 import std.traits;
 import core.stdc.stdio;
 import core.time;
 
 mixin AddLogger!();
 
+struct ProxyProtocol
+{
+    /// Protocol v2 signature
+    private immutable ubyte[12] v2_signature = [
+        13, 10, 13, 10, 0, 13, 10, 81, 85, 73, 84, 10
+    ];
+
+    /// Source address
+    Address src;
+
+    /// Destination address
+    Address dst;
+
+    ///
+    public this (TCPConnection stream) @trusted
+    {
+        ubyte[108] buffer;
+
+        stream.read(buffer[0..12]); // Read signature
```

Comment:
Actually, I don't like reading stream data during ctor'ing a struct. But this way we can drop connections early, see `ensure`s below.